ibridges/irodsconnector/data_operations.py

```diff
@@ -110,56 +120,157 @@ def irods_put(self, local_path: str, irods_path: str, resc_name: str = ''):
+        """Download a collection to the local filesystem
+
+        Parameters
+        ----------
+        irods_path : IrodsPath
+            Absolute irods source path pointing to a collection
+        local_path : Path
+            Absolute path to the destination directory
+        overwrite : bool
+            Overwrite existing local data
+        options : dict
+            More options for the download
+        """
+        if not irods_path.is_collection():
+            raise ValueError("irods_path must be a collection.")
+
+        # get all data objects
+        coll = self.get_collection(irods_path)
+        all_objs = self._get_data_objects(coll)
+
+        for o in all_objs:
+            # ensure local folder exists
+            dest = Path(local_path, o[0].removeprefix(str(irods_path)).lstrip('/'))
+            if not dest.is_dir():
+                os.makedirs(dest)
+            self.irods_get(IrodsPath(self.session, o[0], o[1]), local_path, overwrite, options)
+
+    def get_size(self,
+                 item: Union[irods.data_object.iRODSDataObject, irods.collection.iRODSCollection]) -> int:
+        """Collect the sizes of a data object or a
+        collection.
+
+        Parameters
+        ----------
+        item : iRODSDataObject or iRODSCollection
+
+
+        Returns
+        -------
+        int
+            Total size [bytes] of the iRODS object or all iRODS objects in the collection.
+
+        """
+        if self.has_type_dataobj(item):
+            return item.size
+        elif self.has_type_collection(item):
+            all_objs = self._get_data_objects(item)
+            return sum([size for _, _, size, _ in all_objs])
+        else:
+            raise ValueError("Item must be an iRODS object or iRODS collection.")
+
+    def _get_data_objects(self, coll: irods.collection.iRODSCollection) -> list[str, str, int, str]:
+        """Retrieve all data objects in a collection and all its subcollections.
+
+        Parameters
+        ----------
+        coll : irods.collection.iRODSCollection
+            The collection to search for all data pbjects
+
+        Returns
+        -------
+        list of all data objects
+            [(cllection path, name, size, checksum)]
+        """
+
+        # all objects in the collection
+        objs = [(obj.collection.path, obj.name, obj.size, obj.checksum) for obj in coll.data_objects]
+
+        # all objects in subcollections
+        data_query = self.session.irods_session.query(kw.COLL_NAME, kw.DATA_NAME,
+                                                      DataObject.size, DataObject.checksum)
+        data_query = data_query.filter(kw.LIKE(kw.COLL_NAME, coll.path+"/%"))
+        for res in data_query.get_results():
+            path, name, size, checksum = res.values()
+            objs.append((path, name, size, checksum))
+
+        return objs
+
+    def create_collection(self, coll_path: Union[IrodsPath, str]) -> irods.collection.iRODSCollection:
```

Comment:
`create_collection()` conveniently accepts both `str` and `IrodsPath`, maybe something similar can be implemented for the `get`- and `put`-functions?

Reply:
I adjusted that to all functions